website/blog/2024/09-30-jsdom-based-scraping/index.md

```diff
@@ -0,0 +1,262 @@
+
+:::note
+
+Before going ahead, I want to mention that Crawlee natively supports JSDOM using the [JSDOM Crawler](https://crawlee.dev/api/jsdom-crawler). It gives a framework for the parallel crawling of web pages using plain HTTP requests and jsdom DOM implementation. It uses raw HTTP requests to download web pages, it is very fast and efficient on data bandwidth.
+
+:::
+
+Let’s see how we are going to create the `getApiUrlWithVerificationToken` function:
+
+```js
+const getApiUrlWithVerificationToken = async (body, url) => {
+    log.info(`Getting API session`);
+    const virtualConsole = new VirtualConsole();
+    const { window } = new JSDOM(body, {
+        url,
+        contentType: 'text/html',
+        runScripts: 'dangerously',
+        resources: 'usable' || new CustomResourceLoader(),
+        // ^ 'usable' faster than custom and works without canvas
+        pretendToBeVisual: false,
+        virtualConsole,
+ });
+    virtualConsole.on('error', () => {
+        // ignore errors cause by fake XMLHttpRequest
+ });
+
+    const apiHeaderKeys = ['anonymous-user-id', 'timestamp', 'user-sign'];
+    const apiValues = {};
+    let retries = 10;
+    // api calls made outside of fetch, hack below is to get URL without actual call
+    window.XMLHttpRequest.prototype.setRequestHeader = (name, value) => {
+        if (apiHeaderKeys.includes(name)) {
+            apiValues[name] = value;
+ }
+        if (Object.values(apiValues).length === apiHeaderKeys.length) {
+            retries = 0;
+ }
+ };
+    window.XMLHttpRequest.prototype.open = (method, urlToOpen) => {
+        if (
+ ['static', 'scontent'].find((x) =>
+                urlToOpen.startsWith(`https://${x}`),
+ )
+ ) {
+ }
+        log.debug('urlToOpen', urlToOpen);
+ };
+    do {
+        await sleep(4000);
+        retries--;
+ } while (retries > 0);
+
+    await window.close();
+    return apiValues;
+};
+```
+
+In this function, we are creating a virtual console that uses `CustomResourceLoader` to run the background process and replace the browser with JSDOM.
+
+For this particular example, we need three mandatory headers to make the API call, and those are `anonymous-user-id,` `timestamp,` and `user-sign.`
+
+Using `XMLHttpRequest.prototype.setRequestHeader`, we are checking if the mentioned headers are in the response or not, if yeas, we take the value of those headers, and repeat the retries until we get all the headers.
+
+Then, the most important part is that we use `XMLHttpRequest.prototype.open` to extract the auth data and make calls without actually using browsers or exposing the bot activity.
+
+At the end of `createSessionFunction`, it returns a session with the required headers.
+
+Now coming to our main code, we will use CheerioCrawler and will use `prenavigationHooks` to inject the headers that we got from the earlier function into the `requestHandler`.
+
+```js
+const crawler = new CheerioCrawler({
```

Comment:
ok, i see you broke all the code blocks actually :D this one is also misformatted.

Reply:
Grammarly it is 🤦  updating

Reply:
ouch, formatting code blocks with grammarly, that's bold :D 

i can format this myself if you'd struggle with that

Reply:
was editing english with it :)

Reply:
done using prettier now.